cvat/apps/engine/rq.py

```diff
@@ -0,0 +1,527 @@
+    )
+    exc_type = MutableRQMetaAttribute(
+        RQJobMetaField.EXCEPTION_TYPE,
+        validator=lambda x: issubclass(x, BaseException),
+        optional=True,
+    )
+    exc_args = MutableRQMetaAttribute(
+        RQJobMetaField.EXCEPTION_ARGS,
+        validator=lambda x: isinstance(x, tuple),
+        optional=True,
+    )
+
+    @staticmethod
+    def _get_resettable_fields() -> list[RQJobMetaField]:
+        return [
+            RQJobMetaField.FORMATTED_EXCEPTION,
+            RQJobMetaField.EXCEPTION_TYPE,
+            RQJobMetaField.EXCEPTION_ARGS,
+        ]
+
+
+class BaseRQMeta(RQMetaWithFailureInfo):
+    # immutable && required fields
+    user: UserMeta = UserRQMetaAttribute()
+    request: RequestMeta = RequestRQMetaAttribute()
+
+    # immutable && optional fields
+    org_id: int | None = ImmutableRQMetaAttribute(
+        RQJobMetaField.ORG_ID, validator=lambda x: isinstance(x, int), optional=True
+    )
+    org_slug: int | None = ImmutableRQMetaAttribute(
+        RQJobMetaField.ORG_SLUG, validator=lambda x: isinstance(x, str), optional=True
+    )
+    project_id: int | None = ImmutableRQMetaAttribute(
+        RQJobMetaField.PROJECT_ID, validator=lambda x: isinstance(x, int), optional=True
+    )
+    task_id: int | None = ImmutableRQMetaAttribute(
+        RQJobMetaField.TASK_ID, validator=lambda x: isinstance(x, int), optional=True
+    )
+    job_id: int | None = ImmutableRQMetaAttribute(
+        RQJobMetaField.JOB_ID, validator=lambda x: isinstance(x, int), optional=True
+    )
+
+    # mutable && optional fields
+    progress: float | None = MutableRQMetaAttribute(
+        RQJobMetaField.PROGRESS, validator=lambda x: isinstance(x, float), optional=True
+    )
+    status: str | None = MutableRQMetaAttribute(
+        RQJobMetaField.STATUS, validator=lambda x: isinstance(x, str), optional=True
+    )
+
+    @staticmethod
+    def _get_resettable_fields() -> list[RQJobMetaField]:
+        return RQMetaWithFailureInfo._get_resettable_fields() + [
+            RQJobMetaField.PROGRESS,
+            RQJobMetaField.STATUS,
+        ]
+
+    @classmethod
+    def build(
+        cls,
+        *,
+        request: ExtendedRequest,
+        db_obj: Model | None,
+    ):
+        # to prevent circular import
+        from cvat.apps.events.handlers import job_id, organization_slug, task_id
+        from cvat.apps.webhooks.signals import organization_id, project_id
+
+        oid = organization_id(db_obj)
+        oslug = organization_slug(db_obj)
+        pid = project_id(db_obj)
+        tid = task_id(db_obj)
+        jid = job_id(db_obj)
+
+        user = request.user
+
+        return cls.for_meta(
+            {
+                RQJobMetaField.USER: UserMeta(
+                    {
+                        RQJobMetaField.UserField.ID: user.id,
+                        RQJobMetaField.UserField.USERNAME: user.username,
+                        RQJobMetaField.UserField.EMAIL: getattr(user, "email", ""),
```

Comment:
How could `email` be a missing attribute? Isn't it part of the `User` model?

Reply:
fixed